database/redis/metric.go

```diff
@@ -136,40 +137,87 @@ func (connector *DbConnector) SaveMetrics(metrics map[string]*moira.MatchedMetri
 			}
 
 			var metricEventsChannel = metricEventsChannels[rand.Intn(len(metricEventsChannels))]
-			if err = c.Publish(ctx, metricEventsChannel, event).Err(); err != nil {
-				return err
-			}
+			pipe.SAdd(ctx, metricEventsChannel, event)
 		}
 	}
+
+	if _, err = pipe.Exec(ctx); err != nil {
+		connector.logger.Errorf("Sending metric event error. Error: %v;", err)
+		return err
+	}
+
 	return nil
 }
 
 // SubscribeMetricEvents creates subscription for new metrics and return channel for this events
-func (connector *DbConnector) SubscribeMetricEvents(tomb *tomb.Tomb) (<-chan *moira.MetricEvent, error) {
-	metricsChannel := make(chan *moira.MetricEvent, pubSubWorkerChannelSize)
-	dataChannel, err := connector.manageSubscriptions(tomb, metricEventsChannels)
-	if err != nil {
+func (connector *DbConnector) SubscribeMetricEvents(tomb *tomb.Tomb, params *moira.SubscribeMetricEventsParams) (<-chan *moira.MetricEvent, error) {
+	responesChannel := make(chan string, pubSubWorkerChannelSize)
+	metricChannel := make(chan *moira.MetricEvent, pubSubWorkerChannelSize)
+
+	ctx := connector.context
+	c := *connector.client
+
+	if err := c.Ping(ctx).Err(); err != nil {
 		return nil, err
 	}
 
+	go func() {
+		<-tomb.Dying()
+		close(responesChannel)
+	}()
+
 	go func() {
 		for {
-			data, ok := <-dataChannel
+			response, ok := <-responesChannel
 			if !ok {
-				connector.logger.Info("No more subscriptions, channel is closed. Stop process data...")
-				close(metricsChannel)
+				close(metricChannel)
 				return
 			}
+
 			metricEvent := &moira.MetricEvent{}
-			if err := json.Unmarshal(data, metricEvent); err != nil {
-				connector.logger.Errorf("Failed to parse MetricEvent: %s, error : %v", string(data), err)
+			if err := json.Unmarshal([]byte(response), metricEvent); err != nil {
+				connector.logger.Errorf("Failed to parse MetricEvent: %s, error : %v", response, err)
 				continue
 			}
-			metricsChannel <- metricEvent
+			metricChannel <- metricEvent
 		}
 	}()
 
-	return metricsChannel, nil
+	for channelIdx := 0; channelIdx < len(metricEventsChannels); channelIdx++ {
+		metricEventsChannel := metricEventsChannels[channelIdx]
+		go func() {
+			var popDelay time.Duration
+			for {
+				startPop := time.After(popDelay)
+				select {
+				case <-tomb.Dying():
+					return
+				case <-startPop:
+					data, err := c.SPopN(ctx, metricEventsChannel, params.BatchSize).Result()
+					popDelay = connector.handlePopResponse(data, err, responesChannel, params.Delay)
+				}
+			}
+		}()
+	}
+
+	return metricChannel, nil
+}
+
+func (connector *DbConnector) handlePopResponse(data []string, popError error, responseChannel chan string, defaultDelay time.Duration) time.Duration {
+	if popError != nil {
+		if popError != redis.Nil {
+			connector.logger.Errorf("Failed to pop new metric events. Error: %s.", popError)
+		}
+		return receiveErrorSleepDuration
```

Comment:
Would use please explain, why we need to have specific delay on error and specific delay on empty? I think default delay is enough, but I can mistake

Reply:
The separation can help to adjust program better.

Reply:
Yes, I understand it. But the question was about problem, not about advantage of solution